consensus/istanbul/backend/announce.go

```diff
@@ -17,405 +17,671 @@
+	// If both conditions are true, then this node should announce.
+	checkIfShouldAnnounceTicker := time.NewTicker(5 * time.Second)
+
+	// Create a ticker to check peers' announce versions one every 10 minutes
+	announceVersionsCheckTicker := time.NewTicker(10 * time.Minute)
+
+	// Create all the variables needed for the periodic gossip
+	var announceGossipTicker *time.Ticker
+	var announceGossipTickerCh <-chan time.Time
+	var announceGossipFrequencyState AnnounceGossipFrequencyState
+	var currentAnnounceGossipTickerDuration time.Duration
+	var numGossipedMsgsInHighFreqAfterFirstPeerState int
+
+	for {
+		select {
+		case <-checkIfShouldAnnounceTicker.C:
+			logger.Trace("Checking if this node should announce it's enode")
+
+			shouldAnnounce, err := sb.shouldGenerateAndProcessAnnounce()
+			if err != nil {
+				logger.Warn("Error in checking if should announce", err)
+				break
+			}
+
+			if shouldAnnounce && announceGossipTickerCh == nil {
+				// Immediately gossip an announce
+				go sb.gossipAnnounce()
+
+				if sb.config.AnnounceAggressiveGossipOnEnablement {
+					announceGossipFrequencyState = HighFreqBeforeFirstPeerState
 
-func (ar *announceRecord) String() string {
-	return fmt.Sprintf("{DestAddress: %s, EncryptedEnodeURL length: %d}", ar.DestAddress.String(), len(ar.EncryptedEnodeURL))
+					// Send an announce message once a minute
+					currentAnnounceGossipTickerDuration = 1 * time.Minute
+
+					numGossipedMsgsInHighFreqAfterFirstPeerState = 0
+				} else {
+					announceGossipFrequencyState = LowFreqState
+					currentAnnounceGossipTickerDuration = time.Duration(sb.config.AnnounceGossipPeriod) * time.Second
+				}
+
+				// Enable periodic gossiping by setting announceGossipTickerCh to non nil value
+				announceGossipTicker = time.NewTicker(currentAnnounceGossipTickerDuration)
+				announceGossipTickerCh = announceGossipTicker.C
+				logger.Trace("Enabled periodic gossiping of announce message")
+			} else if !shouldAnnounce && announceGossipTickerCh != nil {
+				// Disable periodic gossiping by setting announceGossipTickerCh to nil
+				announceGossipTicker.Stop()
+				announceGossipTickerCh = nil
+				logger.Trace("Disabled periodic gossiping of announce message")
+			}
+
+			// Use this timer to also prune all announce related data structures.
+			if err := sb.pruneAnnounceDataStructures(); err != nil {
+				logger.Warn("Error is pruning announce data structures", "err", err)
+			}
+
+		case <-announceGossipTickerCh: // If this is nil (when shouldAnnounce was most recently false), this channel will never receive an event
+			logger.Trace("Going to gossip an announce message", "announceGossipFrequencyState", announceGossipFrequencyState, "numGossipedMsgsInHighFreqAfterFirstPeerState", numGossipedMsgsInHighFreqAfterFirstPeerState)
+			switch announceGossipFrequencyState {
+			case HighFreqBeforeFirstPeerState:
+				if len(sb.broadcaster.FindPeers(nil, p2p.AnyPurpose)) > 0 {
```

Comment:
What happens when `sb.gossipAnnounce` is called when `len(sb.broadcaster.FindPeers(nil, p2p.AnyPurpose)) == 0`. Based on a quick reading, it generates and caches the message, then doesn't send it, right? Is there any reason we'd want to do that?

Reply:
The reason it caches is that when peers connect to it, this node will send to them it's cached entry (and then those nodes will subsequently gossip it out to to the rest of the network).

Reply:
👍 That makes sense.